processor/src/main/java/org/mapstruct/ap/internal/model/BuilderFinisherMethodResolver.java

```diff
@@ -0,0 +1,109 @@
+/**
+ *  Copyright 2012-2017 Gunnar Morling (http://www.gunnarmorling.de/)
+ *  and/or other contributors as indicated by the @authors tag. See the
+ *  copyright.txt file in the distribution for a full listing of all
+ *  contributors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.mapstruct.ap.internal.model;
+
+import java.util.Collection;
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.type.DeclaredType;
+
+import org.mapstruct.ap.internal.model.common.BuilderType;
+import org.mapstruct.ap.internal.model.source.Method;
+import org.mapstruct.ap.internal.prism.BuilderMappingPrism;
+import org.mapstruct.ap.internal.util.MapperConfiguration;
+import org.mapstruct.ap.internal.util.Message;
+import org.mapstruct.ap.internal.util.Strings;
+
+import static org.mapstruct.ap.internal.util.Collections.first;
+
+/**
+ * @author Filip Hrisafov
+ */
+public class BuilderFinisherMethodResolver {
+
+    private BuilderFinisherMethodResolver() {
+    }
+
+    public static MethodReference getBuilderFinisherMethod(Method method, BuilderType builderType,
+        MappingBuilderContext ctx) {
+        Collection<ExecutableElement> buildMethods = builderType.getBuildMethods();
+        if ( buildMethods.isEmpty() ) {
+            //If we reach this method this should never happen
+            return null;
+        }
+        else if ( buildMethods.size() == 1 ) {
```

Comment:
If a `@BuilderMapping` was defined with an explicit name for the `builderFinisher`, but only a different method was found, the wrong method will be used. The user would not notice this.

In case he entered a `builderFinisher` I would expect a warning or even better an error in case the only method found has a different name. The issue is most likely how we should know that he explicitly entered  a name, because you defined `build` as a default name in the annotation. So we will always have a `builderFinisher` as soon as the `@BuilderMapping` annotation _(or whatever it will be at the end)_ will be used.

The only way I see to solve this issue is to use an empty string as a default annotation attribute value and in this case use `build` as a constant value here in the code (the `DEFAULT_BUILD_METHOD_NAME`).
And switch the `if-else` ordering.. if a `builderFinisher` was defined try to use a method with exactly this name, if not possible -> error. If nothing was defined fallback to the way you are doing it right now (`size == 1`, then `methodName equals build`).